### ISSUE TITLE
Discussion: make DataFile Serializable && Deserializable

### ISSUE THREAD

**Issue by ZENOTME** [apache/iceberg-rust#774]
## Context 

Make Datafile Serializable && Deserializable is useful, e.g. In distributed compute engine, it will create multiple writers in multiple machines and write the data in parallel and get the DataFile as the results, these DataFiles will be sent to a coordinator and append using transaction. In this case, DataFile should able to be Serializable && Deserializable.

## Solution 

For now, we support Serialize DataFile in _serde module and we should convert the DataFile to _serde::DataFile first, the interface looks like:  `pub fn try_from(value: super::DataFile, partition_type: &StructType,is_version_1: bool) -> _serde::DataFile`. More detail: https://github.com/apache/iceberg-rust/blob/98cd34dc03cd87b330c7bff8fe9f3241746062ac/crates/iceberg/src/spec/manifest.rs#L1361.

There is something we need to resolve to support Datafile Serializable && Deserializable:
1. The related interface needs to be exposed to the public 
2. The interface is not friendly. If the DataFile can be self-contain, things will be easier, e.g. DataFile itself can be Serialize && Deserialize, the user doesn't need to convert it using an interface like `pub fn try_from(value: super::DataFile, partition_type: &StructType,is_version_1: bool) -> _serde::DataFile`

To solve the above, I think there are two solutions:
1. Make DateFile self-contain, store the partition type and version in DataFile directly so that it converts into _serde::DataFile directly and it can be Serialize && Deserialize.
2. Provide something like 
```
struct SerializableDataFile {
  version: i32,
  partition_type: StructType
  data_file: DataFile
}
```
I prefer solution 1 because it looks more natural. Welcome to different opinions and solutions. cc @liurenjie1024 @Fokko @Xuanwo @c-thiel

**Comment by Xuanwo** [comment]
Hi, thank you @ZENOTME for starting this discussion. I prefer to make `DataFile` itself serializable. Maybe we can do this samething for `DataFile` like `TableMetadata`: `TableMetadata` itself is `Serialize` but we won't expose the underlying types.

**Comment by ZENOTME** [comment]
> Hi, thank you @ZENOTME for starting this discussion. I prefer to make `DataFile` itself serializable. Maybe we can do this samething for `DataFile` like `TableMetadata`: `TableMetadata` itself is `Serialize` but we won't expose the underlying types.

To make `DataFile` itself serializable, I think we should contain the struct type of DataFile in it.

**Comment by ZENOTME** [comment]
> Thanks @ZENOTME for raising this. In your question, the typical use case is compute engine, which needs to serialized `DataFile`. But in fact, what a compute engines required is to serialize/deserialized planned task if I understad correctly. If there are use cases `DataFile` itself must be serialized/deserialized, I prefer solution 1.

For read part, the compute engines required is to serialize/deserialized planned task. But for write part,  the compute engines required to pass the data file. #797 is implement used solution 1.

**Comment by ZENOTME** [comment]
Yes, I think https://github.com/apache/iceberg-rust/pull/797 is enough now.
